tests/functional/pykmip/ClusterClient.spec.js

```diff
@@ -0,0 +1,508 @@
+                            reject(err);
+                        } else {
+                            resolve(child);
+                        }
+                    }, 10);
+                });
+                child.on('error', err => {
+                    if (spawnTimeout) {
+                        clearTimeout(spawnTimeout);
+                    }
+                    reject(`${err.toString()}\nStderr: ${stderr}`);
+                });
+                child.on('close', (code, signal) => {
+                    if (spawnTimeout) {
+                        clearTimeout(spawnTimeout);
+                    }
+                    reject(new Error(`socat closed with code ${code} and signal ${signal}.\nStderr: ${stderr}`));
+                });
+            });
+        }
+
+        async function stopSocat(socat) {
+            await new Promise((resolve) => {
+                socat.on('close', resolve);
+                socat.kill('SIGKILL');
+            });
+        }
+
+        beforeEach(async () => {
+            socatProcess = await spawnSocat();
+            client = new ClusterClient(clusterOptions);
+            healthyClients = client.clients;
+            unhealthyClients = client.unhealthyClients;
+            promises = promisify(client);
+            spys = clusterSpys(client.clients);
+            clock = sinon.useFakeTimers({ shouldAdvanceTime: true });
+        });
+
+        afterEach(async () => {
+            if (clock) {
+                clock.reset();
+            }
+        });
+
+        it('should succeed healthcheck', async () => {
+            await assert.doesNotReject(promises.healthcheck(logger), showArsenalErr);
+            assertHealthcheck(
+                { actual: healthyClients.length, expected: nb + 1 },
+                { actual: unhealthyClients.length, expected: 0 },
+                spys.healthchecks,
+            );
+        });
+
+        it(`should encrypt x${NB_OF_ENCRYPTION} in round robin on each ${nb + 1} hosts or ${nb} hosts.`
+            + ' With 1 host disconnecting and reconnecting', async () => {
+            await assert.doesNotReject(async () => {
+                kmsKey = await promises.createBucketKey('testBucket', logger);
+            }, showArsenalErr);
+
+            await cipherAndAssert(
+                promises.cipherDataKey.bind(null, 1, kmsKey, dataKey, logger),
+                nb + 1,
+                { healthy: healthyClients, unhealthy: unhealthyClients },
+                { healthy: nb + 1, unhealthy: 0 },
+                { spysCipherHealthy: spys.cipherDataKeys },
+                false,
+            );
+
+            spys.cipherDataKeys.forEach(spy => spy.resetHistory());
+
+            await stopSocat(socatProcess);
+
+            // extract unhealthy spy from healthy
+            const spysCipherUnhealthy = spys.cipherDataKeys.splice(spys.cipherDataKeys.length - 1, 1);
```

Comment:
Is it always true that always last client is the one that fails

Reply:
in that test yes because I pushed the client that I use to simulate disconnection at the end of the transport list